src/dagParser.cpp

```diff
@@ -0,0 +1,152 @@
+
+    // Communication task with optional payload size
+    if (nodeJson.contains("communicationTask") &&
+        nodeJson["communicationTask"].contains("payloadSize")) {
+      nodeInstance.communicationTask.payloadSize =
+          nodeJson["communicationTask"]["payloadSize"].get<uint32_t>();
+    }
+
+    // Optional downstream nodes
+    if (nodeJson.contains("downstream")) {
+      std::vector<Edge> edges;
+      for (const auto &edgeJson : nodeJson.at("downstream")) {
+        Edge edgeInstance;
+        edgeInstance.id = edgeJson.at("id").get<std::string>();
+        edgeInstance.target = edgeJson.at("target").get<std::string>();
+        edgeInstance.max_partitions =
+            edgeJson.at("max_partitions").get<uint32_t>();
+        edges.push_back(edgeInstance);
+      }
+      nodeInstance.downstream = edges;
+    }
+
+    nodeList.push_back(nodeInstance);
+  }
+
+  return DAGParser(appName, nodeList);
+}
+
+void DAGParser::printNodeDetails() {
+  std::cout << "Application Name: " << applicationName << std::endl;
+
+  for (const auto &node : nodes) {
+    std::cout << "Task: " << node.task << ", Name: " << node.name << std::endl;
+    std::cout << "  Description: " << node.description << std::endl;
+    std::cout << "  Executor: " << node.executor << std::endl;
+    std::cout << "  Container Image: " << node.container.image << std::endl;
+    std::cout << "  Resources - CPU: " << node.container.resources.cpu
+              << ", Memory: " << node.container.resources.memory << std::endl;
+    std::cout << "  Task Complexity: " << node.scalingParameters.taskComplexity
+              << std::endl;
+
+    if (node.communicationTask.payloadSize.has_value()) {
+      std::cout << "  Payload Size: "
+                << node.communicationTask.payloadSize.value() << std::endl;
+    }
+
+    if (node.downstream.has_value()) {
+      std::cout << "  Downstream: " << std::endl;
+      for (const auto &edge : node.downstream.value()) {
+        std::cout << "    Edge ID: " << edge.id << ", Target: " << edge.target
+                  << ", Max Partitions: " << edge.max_partitions << std::endl;
+      }
+    }
+  }
+}
+
+const std::vector<Node> &DAGParser::getNodes() { return nodes; }
+
+const Node &DAGParser::findNodeByName(const std::string &nodeName) {
+  auto it =
+      std::find_if(nodes.begin(), nodes.end(), [&nodeName](const Node &node) {
+        return node.name == nodeName;
+      });
+
+  if (it != nodes.end()) {
+    return *it;
+  } else {
+    throw std::runtime_error("Node with name '" + nodeName + "' not found");
+  }
+}
+
+const Edge &DAGParser::findEdgeByName(const std::string &edgeId) {
+  for (const auto &node : nodes) {
+    if (node.downstream.has_value()) {
+      auto it = std::find_if(
+          node.downstream->begin(), node.downstream->end(),
+          [&edgeId](const Edge &edge) { return edge.id == edgeId; });
+
+      if (it != node.downstream->end()) {
+        return *it;
+      }
+    }
```

Comment:
Minor refactoring, getting rid of one level of nesting:

```suggestion
    if (!node.downstream.has_value()) {
       continue;
    }
    
    auto it = std::find_if(
        node.downstream->begin(), node.downstream->end(),
        [&edgeId](const Edge &edge) { return edge.id == edgeId; });

    if (it != node.downstream->end()) {
      return *it;
    }
```